File: eunseob/w6_dfs/24481.py
import sys

print = sys.stdout.write
input = sys.stdin.readline
sys.setrecursionlimit(100000000)


N, M, R = map(int, input().split())
visited = [False for _ in range(N + 1)]
graph = [list() for _ in range(N + 1)]
depth = [-1 for _ in range(N + 1)]

for _ in range(M):
    s, e = map(int, input().split())
    graph[s].append(e)
    graph[e].append(s)

for g in graph:
    g.sort()


def dfs(d, R):
    visited[R] = True
    depth[R] = d
    for x in graph[R]:
        if not visited[x]:
            dfs(d + 1, x)


dfs(0, R)

for result in range(1, N + 1):
    print(str(depth[result]) + "\n")

File: eojin/24481.js
let visited = [];
let depth = [];

const dfs = (e, r, parentDepth) => {
  visited[r] = true;
  depth[r] = parentDepth + 1;

  e.get(r).forEach(x => {
    if(!visited[x]) {
      dfs(e,x, depth[r]);
    }
  });
}

const findDepth = (input) => {
  const info = input[0].split(' ');
  const nodeNum = Number(info[0]);
  const edgeNum = Number(info[1]);
  const startNode = Number(info[2]);

  visited = Array(nodeNum + 1).fill(false);
  depth = Array(nodeNum + 1).fill(-1);

  const edges = new Map();
  for(let i = 0; i < edgeNum; i++) {
    const edge = input[i + 1].split(' ');
    const key = Number(edge[0]);
    const value = Number(edge[1]);
    
    const keyArray = edges.get(key) || [];
    const valueArray = edges.get(value) || [];
    keyArray.push(value);
    valueArray.push(key);

    edges.set(key, keyArray);
    edges.set(value, valueArray);
  }
  for(item of edges.keys()) {
    const values = edges.get(item).sort((a, b) => a - b);
    edges.set(item, values);
  }

  if (edges.has(startNode)) {
    dfs(edges, startNode, -1);
  } else {
    visited[startNode] = true;
    depth[startNode] = 0;
  }
  
  const result = depth.slice(1);
  console.log(result.join('\n'));
}

const fs = require('fs');
let input = fs.readFileSync('/dev/stdin').toString().split('\n');
// let input = fs.readFileSync('text.txt').toString().split('\n');
findDepth(input);